cypress/helpers/util.ts

```diff
@@ -0,0 +1,130 @@
+import type { MermaidConfig } from '../../packages/mermaid/src/config.type.js';
+
+type CypressConfig = {
+  listUrl?: boolean;
+  listId?: string;
+  name?: string;
+};
+type CypressMermaidConfig = MermaidConfig & CypressConfig;
+
+interface CodeObject {
+  code: string;
+  mermaid: CypressMermaidConfig;
+}
+
+const utf8ToB64 = (str: string): string => {
+  return window.btoa(decodeURIComponent(encodeURIComponent(str)));
+};
+
+const batchId: string = 'mermaid-batch' + new Date().getTime();
+
+export const mermaidUrl = (
+  graphStr: string,
+  options: CypressMermaidConfig,
+  api: boolean
+): string => {
+  const codeObject: CodeObject = {
+    code: graphStr,
+    mermaid: options,
+  };
+  const objStr: string = JSON.stringify(codeObject);
+  let url = `http://localhost:9000/e2e.html?graph=${utf8ToB64(objStr)}`;
+  if (api) {
+    url = `http://localhost:9000/xss.html?graph=${graphStr}`;
+  }
+
+  if (options.listUrl) {
+    cy.log(options.listId, ' ', url);
+  }
+
+  return url;
+};
+
+export const imgSnapshotTest = (
+  graphStr: string,
+  _options: CypressMermaidConfig = {},
+  api = false,
+  validation?: any
+): void => {
+  cy.log(JSON.stringify(_options));
+  const options: CypressMermaidConfig = Object.assign(_options);
+  if (!options.fontFamily) {
+    options.fontFamily = 'courier';
+  }
+  if (!options.sequence) {
+    options.sequence = {};
+  }
+  if (!options.sequence || (options.sequence && !options.sequence.actorFontFamily)) {
+    options.sequence.actorFontFamily = 'courier';
+  }
+  if (options.sequence && !options.sequence.noteFontFamily) {
+    options.sequence.noteFontFamily = 'courier';
+  }
+  options.sequence.actorFontFamily = 'courier';
+  options.sequence.noteFontFamily = 'courier';
+  options.sequence.messageFontFamily = 'courier';
+  if (options.sequence && !options.sequence.actorFontFamily) {
+    options.sequence.actorFontFamily = 'courier';
+  }
+  if (!options.fontSize) {
+    options.fontSize = 16;
+  }
+
+  const url: string = mermaidUrl(graphStr, options, api);
+  openURLAndVerifyRendering(url, options, validation);
+};
+
+export const urlSnapshotTest = (
+  url: string,
+  _options: CypressMermaidConfig,
+  _api = false,
+  validation?: any
+): void => {
+  const options: CypressMermaidConfig = Object.assign(_options);
+  openURLAndVerifyRendering(url, options, validation);
+};
+
+export const renderGraph = (
+  graphStr: string,
+  options: CypressMermaidConfig = {},
+  api = false
+): void => {
+  const url: string = mermaidUrl(graphStr, options, api);
+  openURLAndVerifyRendering(url, options);
+};
+
+export const openURLAndVerifyRendering = (
+  url: string,
+  options: CypressMermaidConfig,
+  validation?: any
```

Comment:
I'm not sure about `validation`, I haven't see any diagram uses it so I setted it to `any`.

I so one that used it as `boolean` but it doesn't seem right:

https://github.com/mermaid-js/mermaid/blob/d9db2edf5a07e19c3f5a41db0e375ca844a2754e/cypress/integration/other/external-diagrams.spec.js#L7

Reply:
Validation should have the type we can pass into `cy.should`.

https://docs.cypress.io/api/commands/should


eg: `'have.property', 'rendered', true`. 

Maybe it should be an `unknown[]`, which we can pass in like `cy.get('svg').should(...validation);`.

I'm not 100% sure here.

Reply:
It would work fine for basic one-time validation, but I don't think we can pass current custom validations into the function as a parameter. Some diagrams have more than one `.should` or function validation.
I think we need more validation; using `imgSnapshotTest` isn't enough some times, e.g., configure styles (#4527) or diagrams (#4263).
 
I'm suggesting making the parameter an object; the first parameter is a helper function, and the second is an option passed to it. Since each diagram has its own attributes, maybe create a helper function with some `.should` and `expect` that verifies the passed arguments are as expected. For example:
 
```
// cypress/helpers/pie.ts
export const pieHelper: CypressHelperFunc = (svg, options: PieAttributes) => {
   if (options.width !== undefined) {
     expect(svg).to.have.attr('width', options.width);
   }
   ...
}
```
 
We could add their helper method at the top of each spec file, or maybe add it to `cypress/helpers` with the name of the spec itself.

What do you think?

Reply:
That seems like a great idea.

Reply:
It would need more work, so it would be off the PR topic. Let's file an issue, discuss it there, and implement it when we have time.

Reply:
created one here #4614. Let's leave it any for now, when applying those changes we'll, update it.

I'll resolve this conversation.